File: tests/logTest.py
#!/usr/bin/env python

# LSST Data Management System
# Copyright 2014 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.

"""
This tests the logging system in a variety of ways.
"""

import lsst.log as log
import os
import shutil
import sys
import tempfile
import threading
import unittest

class TestLog(unittest.TestCase):

    class StdoutCapture(object):
        """
        Context manager to redirect stdout to a file.
        """

        def __init__(self, filename):
            self.stdout = None
            self.outputFilename = filename

        def __enter__(self):
            self.stdout = os.dup(1)
            os.close(1)
            os.open(self.outputFilename, os.O_WRONLY | os.O_CREAT | os.O_TRUNC)

        def __exit__(self, type, value, traceback):
            if self.stdout is not None:
                os.close(1)
                os.dup(self.stdout)
                os.close(self.stdout)
                self.stdout = None

    def setUp(self):
        """Make a temporary directory and a log file in it."""
        self.tempDir = tempfile.mkdtemp()
        self.outputFilename = os.path.join(self.tempDir, "log.out")
        self.stdout = None

    def tearDown(self):
        """Remove the temporary directory."""
        shutil.rmtree(self.tempDir)

    def configure(self, configuration):
        """
        Create a configuration file in the temporary directory and populate
        it with the provided string.
        """
        log.configure_prop(configuration.format(self.outputFilename))

    def check(self, reference):
        """Compare the log file with the provided reference text."""
        with open(self.outputFilename, 'r') as f:
            lines = [l.split(']')[-1] for l in f.readlines()]
            reflines = [l + "\n" for l in reference.split("\n") if l != ""]
            map(self.assertEqual, lines, reflines)


###############################################################################

    def testDefaultLogger(self):
        """Check the default root logger name."""
        self.assertEqual(log.getDefaultLoggerName(), "")

    def testBasic(self):
        """
        Test basic log output.  Since the default threshold is INFO, the
        TRACE message is not emitted.
        """
        with TestLog.StdoutCapture(self.outputFilename):
            log.configure()
            log.trace("This is TRACE")
            log.info("This is INFO")
            log.debug("This is DEBUG")
            log.warn("This is WARN")
            log.error("This is ERROR")
            log.fatal("This is FATAL")
            log.info("Format %d %g %s", 3, 2.71828, "foo")
        self.check("""
 INFO root null - This is INFO
 DEBUG root null - This is DEBUG
 WARN root null - This is WARN
 ERROR root null - This is ERROR
 FATAL root null - This is FATAL
 INFO root null - Format 3 2.71828 foo
""")

    def testContext(self):
        """Test the log context/component stack."""
        with TestLog.StdoutCapture(self.outputFilename):
            log.configure()
            log.trace("This is TRACE")
            log.info("This is INFO")
            log.debug("This is DEBUG")
            with log.LogContext("component"):
                log.trace("This is TRACE")
                log.info("This is INFO")
                log.debug("This is DEBUG")
            log.trace("This is TRACE 2")
            log.info("This is INFO 2")
            log.debug("This is DEBUG 2")
            with log.LogContext("comp") as ctx:
                log.trace("This is TRACE 3")
                log.info("This is INFO 3")
                log.debug("This is DEBUG 3")
                ctx.setLevel(log.INFO)
                self.assertEqual(ctx.getLevel(), log.INFO)
                self.assert_(ctx.isEnabledFor(log.INFO))
                log.trace("This is TRACE 3a")
                log.info("This is INFO 3a")
                log.debug("This is DEBUG 3a")
                with log.LogContext("subcomp", log.TRACE) as ctx2:
                    self.assertEqual(ctx2.getLevel(), log.TRACE)
                    log.trace("This is TRACE 4")
                    log.info("This is INFO 4")
                    log.debug("This is DEBUG 4")
                log.trace("This is TRACE 5")
                log.info("This is INFO 5")
                log.debug("This is DEBUG 5")

        self.check("""
 INFO root null - This is INFO
 DEBUG root null - This is DEBUG
 INFO component null - This is INFO
 DEBUG component null - This is DEBUG
 INFO root null - This is INFO 2
 DEBUG root null - This is DEBUG 2
 INFO comp null - This is INFO 3
 DEBUG comp null - This is DEBUG 3
 INFO comp null - This is INFO 3a
 TRACE comp.subcomp null - This is TRACE 4
 INFO comp.subcomp null - This is INFO 4
 DEBUG comp.subcomp null - This is DEBUG 4
 INFO comp null - This is INFO 5
""")

    def testPattern(self):
        """
        Test a complex pattern for log messages, including Mapped
        Diagnostic Context (MDC).
        """
        with TestLog.StdoutCapture(self.outputFilename):
            self.configure("""
log4j.rootLogger=DEBUG, CA
log4j.appender.CA=ConsoleAppender
log4j.appender.CA.layout=PatternLayout
log4j.appender.CA.layout.ConversionPattern=%-5p %c %C %M (%F:%L) %l - %m - %X%n
""")
            log.trace("This is TRACE")
            log.info("This is INFO")
            log.debug("This is DEBUG")

            log.MDC("x", 3)
            log.MDC("y", "foo")
            log.MDC("z", TestLog)

            log.trace("This is TRACE 2")
            log.info("This is INFO 2")
            log.debug("This is DEBUG 2")
            log.MDCRemove("z")

            with log.LogContext("component"):
                log.trace("This is TRACE 3")
                log.info("This is INFO 3")
                log.debug("This is DEBUG 3")
                log.MDCRemove("x")
                log.trace("This is TRACE 4")
                log.info("This is INFO 4")
                log.debug("This is DEBUG 4")

            log.trace("This is TRACE 5")
            log.info("This is INFO 5")
            log.debug("This is DEBUG 5")

            log.MDCRemove("y")

        # Use format to make line numbers easier to change.
        self.check("""
INFO  root  testPattern (logTest.py:{0[0]}) logTest.py({0[0]}) - This is INFO - {{}}
DEBUG root  testPattern (logTest.py:{0[1]}) logTest.py({0[1]}) - This is DEBUG - {{}}
INFO  root  testPattern (logTest.py:{0[2]}) logTest.py({0[2]}) - This is INFO 2 - {{{{x,3}}{{y,foo}}{{z,<class '{1}.TestLog'>}}}}
DEBUG root  testPattern (logTest.py:{0[3]}) logTest.py({0[3]}) - This is DEBUG 2 - {{{{x,3}}{{y,foo}}{{z,<class '{1}.TestLog'>}}}}
INFO  component  testPattern (logTest.py:{0[4]}) logTest.py({0[4]}) - This is INFO 3 - {{{{x,3}}{{y,foo}}}}
DEBUG component  testPattern (logTest.py:{0[5]}) logTest.py({0[5]}) - This is DEBUG 3 - {{{{x,3}}{{y,foo}}}}
INFO  component  testPattern (logTest.py:{0[6]}) logTest.py({0[6]}) - This is INFO 4 - {{{{y,foo}}}}
DEBUG component  testPattern (logTest.py:{0[7]}) logTest.py({0[7]}) - This is DEBUG 4 - {{{{y,foo}}}}
INFO  root  testPattern (logTest.py:{0[8]}) logTest.py({0[8]}) - This is INFO 5 - {{{{y,foo}}}}
DEBUG root  testPattern (logTest.py:{0[9]}) logTest.py({0[9]}) - This is DEBUG 5 - {{{{y,foo}}}}
""".format([x + 174 for x in (0, 1, 8, 9, 14, 15, 18, 19, 22, 23)], __name__))


    def testMDCPutPid(self):
        """
        Test add of PID Mapped Diagnostic Context (MDC).
        """
        pid = os.fork()
        try:

            log.MDC("PID", os.getpid())
            self.configure("""
log4j.rootLogger=DEBUG, CA
log4j.appender.CA=ConsoleAppender
log4j.appender.CA.layout=PatternLayout
log4j.appender.CA.layout.ConversionPattern=%-5p PID:%X{{PID}} %c %C %M (%F:%L) %l - %m%n
""")
            self.assertGreaterEqual(pid, 0, "Failed to fork")

            msg = "This is INFO"
            if pid == 0:
                self.tempDir = tempfile.mkdtemp()
                self.outputFilename = os.path.join(self.tempDir, "log-child.out")
                msg += " in child process"
            elif pid > 0:
                child_pid, child_status = os.wait()
                self.assertEqual(child_status, 0, "Child returns incorrect code")
                msg += " in parent process"

            with TestLog.StdoutCapture(self.outputFilename):
                log.info(msg)
                line = 243
        finally:
            log.MDCRemove("PID")

        # Use format to make line numbers easier to change.
        self.check("""
INFO  PID:{1} root  testMDCPutPid (logTest.py:{0}) logTest.py({0}) - {2}
""".format(line, os.getpid(), msg))

        # don't pass other tests in child process
        if pid is 0:
            os._exit(0)

    def testFileAppender(self):
        """Test configuring logging to go to a file."""
        self.configure("""
log4j.rootLogger=DEBUG, FA
log4j.appender.FA=FileAppender
log4j.appender.FA.file={0}
log4j.appender.FA.layout=SimpleLayout
""")
        log.MDC("x", 3)
        with log.LogContext("component"):
            log.trace("This is TRACE")
            log.info("This is INFO")
            log.debug("This is DEBUG")
        log.MDCRemove("x")

        self.check("""
INFO - This is INFO
DEBUG - This is DEBUG
""")

    def testPythonLogging(self):
        """Test logging through the Python logging interface."""
        with TestLog.StdoutCapture(self.outputFilename):
            import logging
            lgr = logging.getLogger()
            lgr.setLevel(logging.INFO)
            lgr.addHandler(log.LogHandler())
            log.configure()
            lgr.info("This is INFO")
            logging.shutdown()

        self.check("""
 INFO root null - This is INFO
""")

    def testMdcInit(self):

        expected_msg = \
            "INFO  - main thread {{MDC_INIT,OK}}\n" + \
            "INFO  - thread 1 {{MDC_INIT,OK}}\n" + \
            "INFO  - thread 2 {{MDC_INIT,OK}}\n"

        with TestLog.StdoutCapture(self.outputFilename):

            self.configure("""
log4j.rootLogger=DEBUG, CA
log4j.appender.CA=ConsoleAppender
log4j.appender.CA.layout=PatternLayout
log4j.appender.CA.layout.ConversionPattern=%-5p - %m %X%n
""")

            fun = lambda : log.MDC("MDC_INIT", "OK")
            log.MDCRegisterInit(fun)

            log.info("main thread")

            thread = threading.Thread(target=lambda: log.info("thread 1"))
            thread.start()
            thread.join()

            thread = threading.Thread(target=lambda: log.info("thread 2"))
            thread.start()
            thread.join()

        self.check(expected_msg)

        log.MDCRemove("MDC_INIT")

    def testLwpID(self):
        """Test log.lwpID() method."""
        lwp1 = log.lwpID()
        lwp2 = log.lwpID()
        pid = os.getpid()

        self.assertEqual(lwp1, lwp2)
        # LWP should be the same as PID in the main thread
        # or it can be a small number on platforms not supporting LWP
        self.assert_(lwp1 == pid or lwp1 < 10)

####################################################################################
def main():
    unittest.main()

if __name__ == "__main__":
    main()

File: tests/logTest.cc
/* 
 * LSST Data Management System
 * Copyright 2014 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
// System headers
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>
#include <stdexcept>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <thread>
#include <vector>

// Third-party headers
#include <boost/format.hpp>

// Local headers
#include "lsst/log/Log.h"

#define BOOST_TEST_MODULE Log_1
#define BOOST_TEST_DYN_LINK
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-variable"
#include "boost/test/unit_test.hpp"
#pragma clang diagnostic pop

// used to stringify line numbers to make some names below unique
#define STR(X) XSTR(X)
#define XSTR(X) #X

/* Compute line numbers in expected debug log messages
 *
 * This enforces consistency of string containing
 * expected log messages in case this source file
 * is modified
 *
 */
#define LOGS_DEBUG_LINENO(message, args) \
    LOGS_DEBUG(message); \
    lineno_helper(__LINE__, args);

/* Compute line numbers in expected info log messages
 */
#define LOGS_INFO_LINENO(message, args) \
    LOGS_INFO(message); \
    lineno_helper(__LINE__, args);

#define MDC_PID_KEY "PID"

struct LogFixture {
    std::string ofName;
    int fd = -1;
    enum Layout_t { LAYOUT_SIMPLE, LAYOUT_PATTERN, LAYOUT_COMPONENT, LAYOUT_MDC };

    LogFixture() {
        ofName = _tmpnam();
    }

    ~LogFixture() {
        _unlink_tmp(ofName, fd);
    }

    std::string _tmpnam()
    {
        // PATH_MAX should be more than enough to hold the temp dir and file name
        char cname[PATH_MAX];
        strncpy(cname, P_tmpdir "/logTest-XXXXXXXXX", sizeof(cname));
        cname[sizeof(cname)-1] = '\0';  // Just in case
        fd = mkstemp(cname);
        if (fd == -1) {
          throw std::runtime_error("Failed to create temporary file.");
        }
        return std::string (cname);
    }

    void _unlink_tmp(std::string& name, int _fd)
    {
        if (_fd != -1) {
            unlink(name.c_str());
            close(_fd);
      }
    }

    void configure(Layout_t layout) {
        std::string config = "log4j.rootLogger=DEBUG, FA\n"
                "log4j.appender.FA=FileAppender\n"
                "log4j.appender.FA.file=" + ofName + "\n";
        switch (layout) {
            case LAYOUT_SIMPLE:
                config += "log4j.appender.FA.layout=SimpleLayout\n";
                break;
            case LAYOUT_PATTERN:
                config += "log4j.appender.FA.layout=PatternLayout\n"
                        "log4j.appender.FA.layout.ConversionPattern=%-5p %c %C %M (%F:%L) %l - %m - %X%n\n";
                break;
            case LAYOUT_COMPONENT:
                config += "log4j.appender.FA.layout=PatternLayout\n"
                        "log4j.appender.FA.layout.ConversionPattern=%-5p %c - %m%n\n";
                break;
            case LAYOUT_MDC:
                config += "log4j.appender.FA.layout=PatternLayout\n"
                        "log4j.appender.FA.layout.ConversionPattern=%-5p - %m %X%n\n";
                break;
        }
        LOG_CONFIG_PROP(config);
    }

    std::string format_range(const std::string& format_string, const std::vector<std::string>& args)
    {
        boost::format f(format_string);
        for (std::vector<std::string>::const_iterator it = args.begin(); it != args.end(); ++it) {
            f % *it;
        }
        return f.str();
    }

    void check(const std::string& expected) {
        std::ifstream t(ofName.c_str());
        std::string received((std::istreambuf_iterator<char>(t)),
                             std::istreambuf_iterator<char>());
        BOOST_CHECK_EQUAL(expected, received);
    }

    void lineno_helper(const long& line, std::vector<std::string>& args) {
        // cast is to work around a compiler problem; remove when upgrading
        args.push_back(std::to_string(static_cast<long long>(line)));
    }

    void pid_log_helper(const std::string& msg, std::vector<std::string>& args) {

        std::stringstream ss;

        configure(LAYOUT_PATTERN);
        // cast is to work around a compiler problem; remove when upgrading
        LOG_MDC(MDC_PID_KEY, std::to_string(static_cast<long long>(getpid())));

        LOGS_INFO_LINENO(msg, args);

        LOG_MDC_REMOVE(MDC_PID_KEY);

        // Add args for building expected log message
        args.push_back(msg);
        ss << getpid();
        args.push_back(ss.str());
    }

};


BOOST_FIXTURE_TEST_CASE(basic_stream, LogFixture) {
    configure(LAYOUT_SIMPLE);
    LOGS_TRACE("This is TRACE");
    LOGS_INFO("This is INFO");
    LOGS_DEBUG("This is DEBUG");
    LOGS_WARN("This is WARN");
    LOGS_ERROR("This is ERROR");
    LOGS_FATAL("This is FATAL");
    LOGS_INFO("Format " << 3 << " " << 2.71828 << " foo c++");
    check("INFO - This is INFO\n"
          "DEBUG - This is DEBUG\n"
          "WARN - This is WARN\n"
          "ERROR - This is ERROR\n"
          "FATAL - This is FATAL\n"
          "INFO - Format 3 2.71828 foo c++\n");
}


BOOST_FIXTURE_TEST_CASE(context_stream, LogFixture) {
    configure(LAYOUT_COMPONENT);

    LOGS_TRACE("This is TRACE");
    LOGS_INFO("This is INFO");
    LOGS_DEBUG("This is DEBUG");
    {
        LOG_CTX context("componentX");
        LOGS_TRACE("This is TRACE 1");
        LOGS_INFO("This is INFO 1");
        LOGS_DEBUG("This is DEBUG 1");
    }
    LOGS_TRACE("This is TRACE 2");
    LOGS_INFO("This is INFO 2");
    LOGS_DEBUG("This is DEBUG 2");
    {
        LOG_CTX context("compY");
        LOGS_TRACE("This is TRACE 3");
        LOGS_INFO("This is INFO 3");
        LOGS_DEBUG("This is DEBUG 3");
        LOG_SET_LVL(LOG_DEFAULT_NAME(), LOG_LVL_INFO);
        BOOST_CHECK_EQUAL(LOG_GET_LVL(LOG_DEFAULT_NAME()),
                          LOG_LVL_INFO);
        LOGS_TRACE("This is TRACE 3a");
        LOGS_INFO("This is INFO 3a");
        LOGS_DEBUG("This is DEBUG 3a");
        {
            LOG_CTX context("subcompZ");
            LOG_SET_LVL(LOG_DEFAULT_NAME(), LOG_LVL_TRACE);
            BOOST_CHECK_EQUAL(LOG_GET_LVL(LOG_DEFAULT_NAME()),
                              LOG_LVL_TRACE);
            LOGS_TRACE("This is TRACE 4");
            LOGS_INFO("This is INFO 4");
            LOGS_DEBUG("This is DEBUG 4");
        }
        LOGS_TRACE("This is TRACE 5");
        LOGS_INFO("This is INFO 5");
        LOGS_DEBUG("This is DEBUG 5");
    }

    check("INFO  root - This is INFO\n"
          "DEBUG root - This is DEBUG\n"
          "INFO  componentX - This is INFO 1\n"
          "DEBUG componentX - This is DEBUG 1\n"
          "INFO  root - This is INFO 2\n"
          "DEBUG root - This is DEBUG 2\n"
          "INFO  compY - This is INFO 3\n"
          "DEBUG compY - This is DEBUG 3\n"
          "INFO  compY - This is INFO 3a\n"
          "TRACE compY.subcompZ - This is TRACE 4\n"
          "INFO  compY.subcompZ - This is INFO 4\n"
          "DEBUG compY.subcompZ - This is DEBUG 4\n"
          "INFO  compY - This is INFO 5\n");
}


BOOST_FIXTURE_TEST_CASE(pattern_stream, LogFixture) {

    std::string expected_msg =
          "INFO  root pattern_stream test_method (tests/logTest.cc:%1%) tests/logTest.cc(%1%) - This is INFO - {}\n"
          "DEBUG root pattern_stream test_method (tests/logTest.cc:%2%) tests/logTest.cc(%2%) - This is DEBUG - {}\n"
          "INFO  root pattern_stream test_method (tests/logTest.cc:%3%) tests/logTest.cc(%3%) - This is INFO 2 - {{x,3}{y,foo}}\n"
          "DEBUG root pattern_stream test_method (tests/logTest.cc:%4%) tests/logTest.cc(%4%) - This is DEBUG 2 - {{x,3}{y,foo}}\n"
          "INFO  component pattern_stream test_method (tests/logTest.cc:%5%) tests/logTest.cc(%5%) - This is INFO 3 - {{x,3}{y,foo}}\n"
          "DEBUG component pattern_stream test_method (tests/logTest.cc:%6%) tests/logTest.cc(%6%) - This is DEBUG 3 - {{x,3}{y,foo}}\n"
          "INFO  component pattern_stream test_method (tests/logTest.cc:%7%) tests/logTest.cc(%7%) - This is INFO 4 - {{y,foo}}\n"
          "DEBUG component pattern_stream test_method (tests/logTest.cc:%8%) tests/logTest.cc(%8%) - This is DEBUG 4 - {{y,foo}}\n"
          "INFO  root pattern_stream test_method (tests/logTest.cc:%9%) tests/logTest.cc(%9%) - This is INFO 5 - {{y,foo}}\n"
          "DEBUG root pattern_stream test_method (tests/logTest.cc:%10%) tests/logTest.cc(%10%) - This is DEBUG 5 - {{y,foo}}\n";
    std::vector<std::string> args;

    configure(LAYOUT_PATTERN);

    LOGS_TRACE("This is TRACE");
    LOGS_INFO_LINENO("This is INFO", args);
    LOGS_DEBUG_LINENO("This is DEBUG", args);

    LOG_MDC("x", "3");
    LOG_MDC("y", "foo");

    LOGS_TRACE("This is TRACE 2");
    LOGS_INFO_LINENO("This is INFO 2", args);
    LOGS_DEBUG_LINENO("This is DEBUG 2", args);
    LOG_MDC_REMOVE("z");

    {
        LOG_CTX context("component");
        LOGS_TRACE("This is TRACE 3");
        LOGS_INFO_LINENO("This is INFO 3", args);
        LOGS_DEBUG_LINENO("This is DEBUG 3", args);
        LOG_MDC_REMOVE("x");
        LOGS_TRACE("This is TRACE 4");
        LOGS_INFO_LINENO("This is INFO 4", args);
        LOGS_DEBUG_LINENO("This is DEBUG 4", args);
    }
    LOGS_TRACE("This is TRACE 5");
    LOGS_INFO_LINENO("This is INFO 5", args);
    LOGS_DEBUG_LINENO("This is DEBUG 5", args);

    LOG_MDC_REMOVE("y");

    expected_msg = format_range(expected_msg, args);

    check(expected_msg);

}

BOOST_FIXTURE_TEST_CASE(MDCPutPid, LogFixture) {

    std::string msg;
    std::string expected_msg = "INFO  root LogFixture pid_log_helper (tests/logTest.cc:%1%) "
                               "tests/logTest.cc(%1%) - %2% - "
                               "{{" MDC_PID_KEY ",%3%}}\n";
    std::vector<std::string> args;
    pid_t pid = fork();

    BOOST_CHECK_MESSAGE(pid >= 0, "fork() failed!");
    if (pid == 0) {
        msg = "This is INFO in child process";

        // create a log file dedicated to child process
        ofName = _tmpnam();

        pid_log_helper(msg, args);
    }
    else if (pid > 0) {
        msg = "This is INFO in parent process";

        pid_log_helper(msg, args);

        int status;

        int w;
        do {
          w = wait(&status);
        } while (w == -1 && errno == EINTR);
        if (w == -1) {
            perror("waitpid");
            exit(EXIT_FAILURE);
        }
        BOOST_CHECK_EQUAL(0, status);


    }
    expected_msg = format_range(expected_msg, args);

    check(expected_msg);

    if (pid == 0) {
        _unlink_tmp(ofName, fd);
        exit(0);
    }

}

BOOST_FIXTURE_TEST_CASE(context1_stream, LogFixture) {
    configure(LAYOUT_COMPONENT);

    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
    LOG_PUSHCTX("component1");
    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
    LOG_PUSHCTX("component2");
    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
    LOG_POPCTX();
    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
    LOG_POPCTX();

    {
        LOG_CTX context1("component3");
        LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
        {
            LOG_CTX context1("component4");
            LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
        }
        LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());
    }
    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());

    // unmatched POP will leave us at root logger
    LOG_POPCTX();
    LOGS_INFO("default logger name is " << LOG_DEFAULT_NAME());

    check("INFO  root - default logger name is \n"
          "INFO  component1 - default logger name is component1\n"
          "INFO  component1.component2 - default logger name is component1.component2\n"
          "INFO  component1 - default logger name is component1\n"
          "INFO  component3 - default logger name is component3\n"
          "INFO  component3.component4 - default logger name is component3.component4\n"
          "INFO  component3 - default logger name is component3\n"
          "INFO  root - default logger name is \n"
          "INFO  root - default logger name is \n");
}

BOOST_FIXTURE_TEST_CASE(context_exc, LogFixture) {
    configure(LAYOUT_COMPONENT);

    // multi-level context will result in exception
    BOOST_CHECK_THROW(LOG_PUSHCTX("x.y"), std::invalid_argument);
}

BOOST_FIXTURE_TEST_CASE(dm_1186, LogFixture) {
    // Test for properly-behaved macro, see DM-1186. MAin point here is that
    // it must compile without error without curly braces after if/else.
    configure(LAYOUT_SIMPLE);

    bool dummy = true;
    if (dummy)
    	LOG_INFO("This is INFO");
    else
    	LOG_WARN("This is WARN");

    check("INFO - This is INFO\n");
}

BOOST_FIXTURE_TEST_CASE(mdc_init, LogFixture) {

    std::string expected_msg =
          "INFO  - main thread {{MDC_INIT,OK}}\n"
          "INFO  - thread 1 {{MDC_INIT,OK}}\n"
          "INFO  - thread 2 {{MDC_INIT,OK}}\n";

    configure(LAYOUT_MDC);

    auto fun = [](){ LOG_MDC("MDC_INIT", "OK"); };
    LOG_MDC_INIT(fun);

    LOGS_INFO("main thread");

    std::thread thread1([]() { LOGS_INFO("thread 1"); });
    thread1.join();

    std::thread thread2([]() { LOGS_INFO("thread 2"); });
    thread2.join();

    check(expected_msg);

    LOG_MDC_REMOVE("MDC_INIT");
}

BOOST_AUTO_TEST_CASE(lwp_id) {

    auto lwp1 = lsst::log::lwpID();
    auto lwp2 = lsst::log::lwpID();
    auto pid = getpid();

    BOOST_CHECK_EQUAL(lwp1, lwp2);
    // LWP should be the same as PID in the main thread
    // or it can be a small number on platforms not supporting LWP
    BOOST_CHECK(lwp1 == pid or lwp1 < 10);
}
